sky/provision/azure/instance.py

```diff
@@ -0,0 +1,215 @@
+    network_client = azure.get_client('network', subscription_id)
+
+    instance = compute_client.virtual_machines.instance_view(
+        resource_group_name=resource_group, vm_name=vm.name).as_dict()
+
+    for status in instance['statuses']:
+        code, state = status['code'].split('/')
+        # skip provisioning status
+        if code == 'PowerState':
+            metadata['status'] = state
+            break
+
+    # get ip data
+    nic_id = vm.network_profile.network_interfaces[0].id
+    metadata['nic_name'] = nic_id.split('/')[-1]
+    nic = network_client.network_interfaces.get(
+        resource_group_name=resource_group,
+        network_interface_name=metadata['nic_name'],
+    )
+    ip_config = nic.ip_configurations[0]
+
+    if not provider_config.get('use_internal_ips', False):
+        public_ip_id = ip_config.public_ip_address.id
+        metadata['public_ip_name'] = public_ip_id.split('/')[-1]
+        public_ip = network_client.public_ip_addresses.get(
+            resource_group_name=resource_group,
+            public_ip_address_name=metadata['public_ip_name'],
+        )
+        metadata['external_ip'] = public_ip.ip_address
+
+    metadata['internal_ip'] = ip_config.private_ip_address
+
+    return metadata
+
+
+def _filter_instances(
+    cluster_name: str,
+    provider_config,
+    tag_filters,
+    worker_only: bool,
+):
+    subscription_id = provider_config['subscription_id']
+    # add cluster name filter to only get nodes from this cluster
+    cluster_tag_filters = {**tag_filters, TAG_RAY_CLUSTER_NAME: cluster_name}
+    if worker_only:
+        cluster_tag_filters[TAG_RAY_NODE_KIND] = 'worker'
+
+    def match_tags(vm):
+        for k, v in cluster_tag_filters.items():
+            if vm.tags.get(k) != v:
+                return False
+        return True
+
+    compute_client = azure.get_client('compute', subscription_id)
+    vms = compute_client.virtual_machines.list(
+        resource_group_name=provider_config['resource_group'])
+
+    nodes = [
+        _extract_metadata(vm, provider_config)
+        for vm in filter(match_tags, vms)
+    ]
+    return {node['name']: node for node in nodes}
+
+
+def stop_instances(
+    cluster_name: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    assert provider_config is not None, cluster_name
+    subscription_id = provider_config['subscription_id']
+    resource_group = provider_config['resource_group']
+
+    instances_to_stop = _filter_instances(
+        cluster_name,
+        provider_config,
+        {},
+        worker_only,
+    )
+    compute_client = azure.get_client('compute', subscription_id)
+
+    stop = get_azure_sdk_function(
+        client=compute_client.virtual_machines,
+        function_name='deallocate',
+    )
+    for node_id in instances_to_stop.keys():
+        stop(resource_group_name=resource_group, vm_name=node_id)
```

Comment:
nit: should we rename `node_id` to `vm_name` or `instane_name`. Reason: we have three concepts here: `instances`, `node` and `vm`. Just wondering if we can make them consistent.